fdbserver/BackupWorker.actor.cpp

```diff
@@ -839,6 +839,10 @@ ACTOR Future<Void> uploadData(BackupData* self) {
 			// make sure file is saved on version boundary
 			popVersion = lastVersion;
 			numMsg = lastVersionIndex;
+
+			// If we aren't able to process any messages and the lock is blocking us from
+			// queuing more, then we are stuck. This could suggest the lock capacity is too small.
+			ASSERT(numMsg > 0 || self->lock->waiters() == 0);
```

Comment:
@jzhou77 It looks like you've written a bunch of this file, so I was hoping your could take a look and see if this ASSERT makes sense to you. The idea is that if we can't pull any messages out of the queue because they are all at the same version, and we can't put new items in the queue because of the lock, then we can't make progress.

Technically, there are some conditions in which we could actually unlock this (e.g. line 862 and 876 both erase messages as special conditions), but it still seems like we shouldn't be allowed to be in the situation described above.

Tracking this bug down took a bit of effort, so my hope is that by having the ASSERT it will be much easier to diagnose if this happens again.

Reply:
Yes, the ASSERT looks good to me.